### PR TITLE
vhost-user-blk device and backend MQ support

### PR DESCRIPTION
**Pull request by yangzhon**
@sboeuf @sameo 

This patch set is for MQ support in the vhost-user-blk device and backend, please help review, thanks.

### CODE REVIEW

**Comment by yangzhon** [comment]
@sboeuf @sameo @rbradford 

There are two panics during the CI building, but i found those issues are not belong to my CI changes. Even i reverted CI patch, those issues still exist.  Could you help me check this? thanks.

cloud login:          Mounting [0;1;39mMount unit for core, revision 7713[0m...
[[0;32m  OK  [0m] Mounted [0;1;39mMount unit for core, revision 7713[0m.
thread 'tests::test_boot_from_vhost_user_blk' panicked at 'assertion failed: guest.get_total_memory().unwrap_or_default() > 492_000', src/main.rs:2810:13

test tests::test_reboot ... ok
thread 'tests::test_boot_from_vhost_user_blk' panicked at 'assertion failed: guest.get_entropy().unwrap_or_default() >= 900', src/main.rs:2811:13

**Comment by rbradford** [comment]
For MQ you need to run multiple threads otherwise you won't get any throughput improvement. Without multiple threads this is effectively making no change.

**Comment by yangzhon** [comment]
> For MQ you need to run multiple threads otherwise you won't get any throughput improvement. Without multiple threads this is effectively making no change.

Justly used 4 queues to check from Qemu + SPDK, and found there is not one thread per one queue in SPDK backend side.

# The SPDK process command
root      3697  199  1.7 135584004 136680 pts/0 RLl Dec17 19487:21 ./app/vhost/vhost -S /var/tmp -s 1024 -m 0x3

#check how many threads from this process
10:57 chen:~$ ps -T -p 3697
  PID  SPID TTY          TIME CMD
 3697  3697 pts/0    6-18:22:43 reactor_0
 3697  3698 pts/0    00:00:00 eal-intr-thread
 3697  3699 pts/0    6-18:22:46 reactor_1
 3697  3704 pts/0    00:00:03 vhost-events

The fio test command:
sudo fio --ioengine=libaio --filename=/dev/vda --direct=1 --rw=randwrite --bs=4k --size=1G --numjobs=4 --runtime=10 --group_reporting --name=file2
sudo fio --ioengine=libaio --filename=/dev/vda --direct=1 --rw=randread --bs=4k --size=1G --numjobs=4 --runtime=10 --group_reporting --name=file2

There are only 4 threads from SPDK whatever the queue numbers is set in qemu command.

**Comment by yangzhon** [comment]
@rbradford ,

I changed the vhost-user-blk backend to direct mode, and used the different queue numbers(1 ,2 ,4) and get almost same test results.
(1).read
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randread --bs=4k --size=1G --numjobs=4 --runtime=10 --group_reporting --name=file2
(2). write
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randwrite --bs=4k --size=1G --numjobs=4 --runtime=10 --group_reporting --name=file2

Test result
(1). rust device + rust backend
read: iops=173
write: iops=460

Thanks for your reminder, i will add the multiple thread to handle MQ, thanks.

**Comment by yangzhon** [comment]
@rbradford @sameo @sboeuf @bjzhjing 

I updated the PR and only add multiple thread into vhost-user-blk device(Notice: i did not add multiple thread into vhost-user-blk backend this time, those changes should be in vhost_user_backend crate and will be done after China new year).

Since the  vhost-user-blk device has been added multiple queues and thread supports, i did vhost-user-blk + SPDK fio test as below:

(1)queue=1
##read
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randread --bs=4k --size=512M --numjobs=1 --runtime=10 --group_reporting --name=file2
**iops=95742**

##write
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randwrite --bs=4k --size=512M --numjobs=1 --runtime=10 --group_reporting --name=file2
**iops=96875**

(2)queue = 4 
##read
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randread --bs=4k --size=512M --numjobs=4 --runtime=10 --group_reporting --name=file2
**iops = 264391**

##write
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randwrite --bs=4k --size=512M --numjobs=4 --runtime=10 --group_reporting --name=file2
**iops=248124**

If patch 4 and 5 are not suitable in this time, patch 1,2 and 3 are okay for SPDK backend. thanks.

**Comment by yangzhon** [comment]
@rbradford @sameo @sboeuf

The MQ and multi-thread for vhost-user-blk device are ready, please help review, thanks!

**Comment by rbradford** [comment]
@bjzhjing Please review @yangzhon's code.

File: vm-virtio/src/vhost_user/blk.rs [hunk 1]
```diff
@@ -106,6 +106,17 @@ impl Blk {
                 .map_err(Error::VhostUserSetProtocolFeatures)?;
         }
 
+        // Get the max queues number from backend, and the queue number set
+        // should be less than this max queue number.
+        let max_queues_num = vhost_user_blk
+            .get_queue_num()
+            .map_err(Error::VhostUserGetQueueMaxNum)?;
+
+        if vu_cfg.num_queues > max_queues_num as usize {
```
Comment (bjzhjing) [inline vm-virtio/src/vhost_user/blk.rs]: Is it not allowed to have vu_cfg.num_queues equal to max_queues_num?

File: src/bin/vhost_user_blk.rs [hunk 2]
```diff
@@ -165,7 +165,7 @@ impl VhostUserBlkBackend {
 
 impl VhostUserBackend for VhostUserBlkBackend {
     fn num_queues(&self) -> usize {
-        NUM_QUEUES
+        MAX_NUM_QUEUES
```
Comment (bjzhjing) [inline src/bin/vhost_user_blk.rs]: Here, the fixed value, MAX_NUM_QUEUES, is specified in num_queues() which is required by VhostUserHandler to initialize vrings, thus, vhost-user-blk will always have 128 vrings initialized, no matter how many vrings vhost-user-blk device actually use. Although only the vrings as many as vhost-user-blk device defined have actual data handling, it's sensible to make the queue numbers flexible to be defined for vhost-user-blk backend.

File: vm-virtio/src/vhost_user/blk.rs (hunk 1 above)
Comment (yangzhon) [reply vm-virtio/src/vhost_user/blk.rs]: vu_cfg.num_queues can be equal to max_queues_num. thanks!

File: src/bin/vhost_user_blk.rs (hunk 2 above)
Comment (yangzhon) [reply src/bin/vhost_user_blk.rs]: This is issue, so i can add the queue_num parameter into backend side. thanks!

**Comment by yangzhon** [comment]
> Please squash [5fdfb94](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/5fdfb9498a6e86a819109470fd7f51a46b6a65bd) with commit [7585a93](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/7585a93f0508ff931b3b37d30bb6cecf28343fab), this should be part of it

squashed in new release

**Comment by bjzhjing** [comment]
The change looks good to me!

**Comment by bjzhjing** [comment]
@rbradford @sboeuf This PR mostly LGTM, the only thing I want to mention is about macro_rules! virtio_pausable in device.rs, the current implementation seems to be the only way to work for all the current cases, since yang and I have tried other ways to make it look simple, but the macro does not recognize any match condition. Have any better suggestion on this? If no, just keep it as is, I approve this PR.

**Comment by sboeuf** [comment]
@yangzhon the PR needs a rebase.

File: vm-virtio/src/device.rs [hunk 3]
```diff
@@ -143,6 +143,36 @@ macro_rules! virtio_pausable_inner {
...
+            );
+            self.paused.store(false, Ordering::SeqCst);
+            if let Some(epoll_thread) = &self.epoll_thread {
+                for i in 0..epoll_thread.len() {
+                    epoll_thread[i].thread().unpark();
+                }
+            }
+            Ok(())
+        }
+    };
```
[the remark this answers is not in the document]
Comment (sameo) [reply vm-virtio/src/device.rs]: I'll have a look at how we could do it differently/better, but that should not prevent this PR from being merged.

**Comment by yangzhon** [comment]
@sameo @rbradford @sboeuf ,

Thanks all for reviewing my PR!
I have rebased the PR and verified it again, there is not any issue.

If this updated PR has not any issue, please merge it, thanks!

File: src/bin/vhost_user_blk.rs [hunk 4]
```diff
@@ -25,8 +25,8 @@ fn main() {
                 .long("block-backend")
                 .help(
                     "vhost-user-block backend parameters \"image=<image_path>,\
-                     sock=<socket_path>,readonly=true|false,\
-                     direct=true|false\"",
+                     sock=<socket_path>,num_queues=<number_of_queues>,\
+                     readonly=true|false,direct=true|false\"",
```
Comment (rbradford) [inline src/bin/vhost_user_blk.rs]: @yangzhon I think you need to update the help for the "cloud-hypervisor" binary too. You only updated the deprecated binary.

Comment (yangzhon) [reply src/bin/vhost_user_blk.rs]: The below one?  
 ```
.arg(
            Arg::with_name("vhost-user-blk")
                .long("vhost-user-blk")
                .help(
                    "Vhost user Block parameters \"sock=<socket_path>,\
                     num_queues=<number_of_queues>,\
                     queue_size=<size_of_each_queue>, \
                     wce=<true|false, default true>\"",
                )
                .takes_value(true)
                .min_values(1)
                .group("vm-config"),
        )
```

The num_queues=<number_of_queues> has been added into cloud-hypervisor binary, although it is only set to 1 before.

Comment (rbradford) [reply src/bin/vhost_user_blk.rs]: @yangzhon https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7cb61d396032aa982e974c36d205aecdd43d67a9/src/main.rs#L286

The vhost_user_blk (and _net) binaries will be removed after 0.5 and the backends run via the cloud-hypervisor binary.

Comment (yangzhon) [reply src/bin/vhost_user_blk.rs]: @rbradford ,

Thanks a lot! I will add a patch for this block-backend. 
By the way, the Google_Summer_of_Code_2019 of Qemu is also implementing one vhost user block backend in the qemu.

**Pull request merged**